cleanlab/count.py

```diff
@@ -43,9 +43,10 @@
     get_num_classes,
     is_torch_dataset,
     is_tensorflow_dataset,
-    int2onehot,
-    _binarize_pred_probs_slice,
+    get_onehot_num_classes,
+    stack_complement,
```

Comment:
can you explain why these two functions can't be in multilabel_utils.py instead?  I'm not sure I understand the circular import issue

Reply:
```ImportError: cannot import name 'get_self_confidence_for_each_label' from partially initialized module 'cleanlab.rank' (most likely due to a circular import)```

Reply:
<img width="955" alt="Screen Shot 2022-10-26 at 12 35 23 PM" src="https://user-images.githubusercontent.com/9433028/197957685-007e43fc-e520-4856-b5d3-9f9bf1b4fd45.png">

Reply:
it works in filter.py, but not in count.py

We should reorganize imports in a separate PR?

Reply:
What about moving a lot of the Scorer-related stuff to a new helper file: cleanlab/internal/multilabel_scorer.py

Eg. could move things like this into there:

```
from cleanlab.rank import (
    get_self_confidence_for_each_label,
    get_normalized_margin_for_each_label,
    get_confidence_weighted_entropy_for_each_label,
)
```

```
class _Wrapper:
    """Helper class for wrapping callable functions as attributes of an Enum instead of
    setting them as methods of the Enum class.
    This class is only intended to be used internally for the ClassLabelScorer or
    other cases where functions are used for enumeration values.
    """

    def __init__(self, f: Callable) -> None:
        self.f = f

    def __call__(self, *args, **kwargs):
        return self.f(*args, **kwargs)

    def __repr__(self):
        return self.f.__name__


class ClassLabelScorer(Enum):
    """Enum for the different methods to compute label quality scores."""

    SELF_CONFIDENCE = _Wrapper(get_self_confidence_for_each_label)
    """Returns the self-confidence label-quality score for each datapoint.
    See also
    --------
    cleanlab.rank.get_self_confidence_for_each_label
    """
    NORMALIZED_MARGIN = _Wrapper(get_normalized_margin_for_each_label)
    """Returns the "normalized margin" label-quality score for each datapoint.
    See also
    --------
    cleanlab.rank.get_normalized_margin_for_each_label
    """
    CONFIDENCE_WEIGHTED_ENTROPY = _Wrapper(get_confidence_weighted_entropy_for_each_label)
    """Returns the "confidence weighted entropy" label-quality score for each datapoint.
    See also
    --------
    cleanlab.rank.get_confidence_weighted_entropy_for_each_label
    """

    def __call__(self, labels: np.ndarray, pred_probs: np.ndarray, **kwargs) -> np.ndarray:
        """Returns the label-quality scores for each datapoint based on the given labels and predicted probabilities.
        See the documentation for each method for more details.
        Example
        -------
        >>> import numpy as np
        >>> from cleanlab.internal.multilabel_utils import ClassLabelScorer
        >>> labels = np.array([0, 0, 0, 1, 1, 1])
        >>> pred_probs = np.array([
        ...     [0.9, 0.1],
        ...     [0.8, 0.2],
        ...     [0.7, 0.3],
        ...     [0.2, 0.8],
        ...     [0.75, 0.25],
        ...     [0.1, 0.9],
        ... ])
        >>> ClassLabelScorer.SELF_CONFIDENCE(labels, pred_probs)
        array([0.9 , 0.8 , 0.7 , 0.8 , 0.25, 0.9 ])
        """
        return self.value(labels, pred_probs, **kwargs)

    @classmethod
    def from_str(cls, method: str) -> "ClassLabelScorer":
        """Constructs an instance of the ClassLabelScorer enum based on the given method name.
        Parameters
        ----------
        method:
            The name of the scoring method to use.
        Returns
        -------
        scorer:
            An instance of the ClassLabelScorer enum.
        Raises
        ------
        ValueError:
            If the given method name is not a valid method name.
            It must be one of the following: "self_confidence", "normalized_margin", or "confidence_weighted_entropy".
        Example
        -------
        >>> from cleanlab.internal.multilabel_utils import ClassLabelScorer
        >>> ClassLabelScorer.from_str("self_confidence")
        <ClassLabelScorer.SELF_CONFIDENCE: get_self_confidence_for_each_label>
        """
        try:
            return cls[method.upper()]
        except KeyError:
            raise ValueError(f"Invalid method name: {method}")


class Aggregator:
    """Helper class for aggregating the label quality scores for each class into a single score for each datapoint.
    Parameters
    ----------
    method:
        The method to compute the label quality scores for each class.
    kwargs:
        Additional keyword arguments to pass to the method when called.
    """

    def __init__(self, method, **kwargs):
        self._validate_method(method)
        self.method = method
        self.kwargs = kwargs

    @staticmethod
    def _validate_method(method) -> None:
        assert callable(method), f"Expected callable method, got {type(method)}"

    @staticmethod
    def _validate_scores(scores: np.ndarray) -> None:
        if not (isinstance(scores, np.ndarray) and scores.ndim == 2):
            raise ValueError(
                f"Expected 2D array for scores, got {type(scores)} with shape {scores.shape}"
            )

    def __call__(self, scores: np.ndarray, **kwargs) -> np.ndarray:
        """Returns the label quality scores for each datapoint based on the given label quality scores for each class.
        Parameters
        ----------
        scores:
            The label quality scores for each class.
        Returns
        -------
        aggregated_scores:
            A single label quality score for each datapoint.
        """
        self._validate_scores(scores)
        kwargs["axis"] = 1
        return self.method(scores, **{**kwargs, **self.kwargs})

    def __repr__(self):
        return f"Aggregator(method={self.method.__name__}, kwargs={self.kwargs})"


def exponential_moving_average(
    s: np.ndarray,
    *,
    alpha: Optional[float] = None,
    axis: int = 1,
    **_,
) -> np.ndarray:
    """Exponential moving average (EMA) score function.
    For a score vector s = (s_1, ..., s_K) with K scores, the values
    are sorted in *descending* order and the exponential moving average
    of the last score is calculated.
    Parameters
    ----------
    s :
        Scores to be transformed.
    alpha :
        Discount factor that determines the weight of the previous EMA score.
        Higher alpha means that the previous EMA score has a lower weight while
        the current score has a higher weight.
        If alpha is None, it is set to 2 / (K + 1) where K is the number of scores.
    axis :
        Axis along which the scores are sorted.
    Returns
    -------
        Exponential moving average score.
    Examples
    --------
    >>> from cleanlab.internal.multilabel_utils import exponential_moving_average
    >>> import numpy as np
    >>> s = np.array([0.1, 0.2, 0.3])
    >>> exponential_moving_average(s, alpha=0.5)
    np.array([0.175])
    """
    K = s.shape[1]
    s_sorted = np.fliplr(np.sort(s, axis=axis))
    if alpha is None:
        # One conventional choice for alpha is 2/(K + 1), where K is the number of periods in the moving average.
        alpha = float(2 / (K + 1))
    s_T = s_sorted.T
    s_ema, s_next = s_T[0], s_T[1:]
    for s_i in s_next:
        s_ema = alpha * s_i + (1 - alpha) * s_ema
    return s_ema


class MultilabelScorer:
    """Aggregates label quality scores across different classes to produce one score per example in multi-label classification tasks.
    Parameters
    ----------
    base_scorer:
        The method to compute the label quality scores for each class.
        See the documentation for the ClassLabelScorer enum for more details.
    aggregator:
        The method to aggregate the label quality scores for each class into a single score for each datapoint.
        Defaults to the EMA (exponential moving average) aggregator with forgetting factor ``alpha=0.8``.
        See the documentation for the Aggregator class for more details.
        See also
        --------
        exponential_moving_average
    strict:
        Flag for performing strict validation of the input data.
    """

    def __init__(
        self,
        base_scorer: ClassLabelScorer = ClassLabelScorer.SELF_CONFIDENCE,
        aggregator: Union[Aggregator, Callable] = Aggregator(exponential_moving_average, alpha=0.8),
        *,
        strict: bool = True,
    ):
        self.base_scorer = base_scorer
        if not isinstance(aggregator, Aggregator):
            self.aggregator = Aggregator(aggregator)
        else:
            self.aggregator = aggregator
        self.strict = strict

    def __call__(self, labels: np.ndarray, pred_probs: np.ndarray, **kwargs) -> np.ndarray:
        """
        Computes a quality score for each label in a multi-label classification problem
        based on out-of-sample predicted probabilities.
        The score is computed by averaging the base_scorer over all labels.
        Parameters
        ----------
        labels:
            A 2D array of shape (n_samples, n_labels) with binary labels.
        pred_probs:
            A 2D array of shape (n_samples, n_labels) with predicted probabilities.
        kwargs:
            Additional keyword arguments to pass to the base_scorer and the aggregator.
        Returns
        -------
        scores:
            A 1D array of shape (n_samples,) with the quality scores for each datapoint.
        Examples
        --------
        >>> from cleanlab.internal.multilabel_utils import MultilabelScorer, ClassLabelScorer
        >>> import numpy as np
        >>> labels = np.array([[0, 1, 0], [1, 0, 1]])
        >>> pred_probs = np.array([[0.1, 0.9, 0.1], [0.4, 0.1, 0.9]])
        >>> scorer = MultilabelScorer()
        >>> scores = scorer(labels, pred_probs)
        >>> scores
        array([0.9, 0.5])
        >>> scorer = MultilabelScorer(
        ...     base_scorer = ClassLabelScorer.NORMALIZED_MARGIN,
        ...     aggregator = np.min,  # Use the "worst" label quality score for each example.
        ... )
        >>> scores = scorer(labels, pred_probs)
        >>> scores
        array([0.9, 0.4])
        """
        if self.strict:
            self._validate_labels_and_pred_probs(labels, pred_probs)
        scores = np.zeros(shape=labels.shape)
        for i, (label_i, pred_prob_i) in enumerate(zip(labels.T, pred_probs.T)):
            pred_prob_i_two_columns = stack_complement(pred_prob_i)
            scores[:, i] = self.base_scorer(label_i, pred_prob_i_two_columns, **kwargs)

        return self.aggregator(scores, **kwargs)

    @staticmethod
    def _validate_labels_and_pred_probs(labels: np.ndarray, pred_probs: np.ndarray) -> None:
        """
        Checks that (multi-)labels are in the proper binary indicator format and that
        they are compatible with the predicted probabilities.
        """
        # Only allow dense matrices for labels for now
        if not isinstance(labels, np.ndarray):
            raise TypeError("Labels must be a numpy array.")
        if not _is_multilabel(labels):
            raise ValueError("Labels must be in multi-label format.")
        if labels.shape != pred_probs.shape:
            raise ValueError("Labels and predicted probabilities must have the same shape.")


def get_label_quality_scores(
    labels, pred_probs, *, method: MultilabelScorer = MultilabelScorer()
) -> np.ndarray:
    """Computes a quality score for each label in a multi-label classification problem
    based on out-of-sample predicted probabilities.
    Parameters
    ----------
    labels:
        A 2D array of shape (N, K) with binary labels.
    pred_probs:
        A 2D array of shape (N, K) with predicted probabilities.
    method:
        A scoring+aggregation method for computing the label quality scores of examples in a multi-label classification setting.
    Returns
    -------
    scores:
        A 1D array of shape (N,) with the quality scores for each datapoint.
    Examples
    --------
    >>> import cleanlab.internal.multilabel_utils as mlutils
    >>> import numpy as np
    >>> labels = np.array([[0, 1, 0], [1, 0, 1]])
    >>> pred_probs = np.array([[0.1, 0.9, 0.1], [0.4, 0.1, 0.9]])
    >>> scores = mlutils.get_label_quality_scores(labels, pred_probs, method=mlutils.MultilabelScorer())
    >>> scores
    array([0.9, 0.5])
    See also
    --------
    MultilabelScorer:
        See the documentation for the MultilabelScorer class for more examples of scoring methods and aggregation methods.
    """
    return method(labels, pred_probs)
```

And then it seems like you could move `get_onehot_num_classes, stack_complement` into multilabel_utils.py and have multilabel_scorer.py import whatever is needed from multilabel_utils.py?

Reply:
If this plan ^ is not that easy for some reason I'm overlooking, let's just keep things as they are.

Reply:
will try that out, thanks!

Reply:
Another idea is to move `get_onehot_num_classes, stack_complement` into their own helper internal file, but these are clearly multilabel specific and should ideally not be in the general utils.py IMO

Reply:
Fixed in commit: https://github.com/cleanlab/cleanlab/pull/509/commits/b6638607e8e92ff82f7cd3f06e4c9fd7e7ec452a